### PR TITLE
Core/Blackmarket: Implemented black market

### PR DESCRIPTION
**Pull request by opcod3**
Implemented the Black Market auction house.
Works as it's supposed to from my testing. Cannot be 100% sure as I have never seen it on retail.

Please give me feedback if anything is wrong or badly implemented so I can fix it and more importantly so I can learn from my mistakes.

### CODE REVIEW

**Comment by Vincent-Michael** [comment]
pls dump your sql with mysqldump

**Comment by Gigi1237** [comment]
@Vincent-Michael will do

File: src/server/game/BlackMarket/BlackMarketMgr.cpp [hunk 1]
```diff
@@ -0,0 +1,488 @@
...
+    return strm.str();
+}
+
+std::string BlackMarketEntry::BuildAuctionMailBody()
+{
+    std::ostringstream strm;
+    strm << GetTemplate()->SellerNPC << ':' << CurrentBid;
+
+    return strm.str();
+}
```
Comment (Vincent-Michael) [inline src/server/game/BlackMarket/BlackMarketMgr.cpp]: fix eof

File: src/server/game/BlackMarket/BlackMarketMgr.h [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+
+      BlackMarketEntryMap mAuctions;
+      BlackMarketTemplateMap mTemplates;
+
+      time_t mLastUpdate;
+};
+
+#define sBlackMarketMgr BlackMarketMgr::instance()
+
+#endif
```
Comment (Vincent-Michael) [inline src/server/game/BlackMarket/BlackMarketMgr.h]: same here

File: src/server/game/BlackMarket/BlackMarketMgr.cpp (hunk 1 above)
Comment (StormBytePP) [inline src/server/game/BlackMarket/BlackMarketMgr.cpp]: No newline at end

File: src/server/game/Handlers/BankHandler.cpp [hunk 3]
```diff
@@ -24,7 +24,7 @@
 
 void WorldSession::HandleAutoBankItemOpcode(WorldPackets::Bank::AutoBankItem& packet)
 {
-    TC_LOG_DEBUG("network", "STORAGE: receive bag = %u, slot = %u", packet.Bag, packet.Slot);
+    TC_LOG_DEBUG("network", "STORAGE: receive b ag = %u, slot = %u", packet.Bag, packet.Slot);
```
Comment (Vincent-Michael) [inline src/server/game/Handlers/BankHandler.cpp]: löl?

File: src/server/game/BlackMarket/BlackMarketMgr.h (hunk 2 above)
Comment (StormBytePP) [inline src/server/game/BlackMarket/BlackMarketMgr.h]: no newline at end

File: src/server/game/BlackMarket/BlackMarketMgr.h [hunk 4]
```diff
@@ -0,0 +1,157 @@
+/*
+* Copyright (C) 2008-2015 TrinityCore <http://www.trinitycore.org/>
+* Copyright (C) 2005-2009 MaNGOS <http://getmangos.com/>
```
Comment (DDuarte) [inline src/server/game/BlackMarket/BlackMarketMgr.h]: I guess this doesn't apply to these files :)

File: src/server/game/Handlers/BankHandler.cpp (hunk 3 above)
Comment (StormBytePP) [inline src/server/game/Handlers/BankHandler.cpp]: Why the space in b ag?

File: src/server/game/Server/Packets/BlackMarketPackets.cpp [hunk 5]
```diff
@@ -30,3 +59,89 @@ WorldPacket const* WorldPackets::BlackMarket::BlackMarketOpenResult::Write()
...
+    data >> blackMarketItem.Quantity;
+    data >> blackMarketItem.MinBid;
+    data >> blackMarketItem.MinIncrement;
+    data >> blackMarketItem.CurrentBid;
+    data >> blackMarketItem.SecondsRemaining;
+    data >> blackMarketItem.NumBids;
+    blackMarketItem.HighBid = data.ReadBit();
+
+    return data;
+}
```
Comment (Vincent-Michael) [inline src/server/game/Server/Packets/BlackMarketPackets.cpp]: same here

File: src/server/game/Handlers/BlackMarketHandler.cpp [hunk 6]
```diff
@@ -41,6 +42,113 @@ void WorldSession::SendBlackMarketOpenResult(ObjectGuid guid, Creature* /*auctio
...
+
+void WorldSession::SendBlackMarketOutbidNotification(BlackMarketTemplate const* templ)
+{
+    WorldPackets::BlackMarket::BlackMarketOutbid packet;
+
+    packet.MarketID = templ->MarketID;
+    packet.Item = templ->Item;
+
+    SendPacket(packet.Write());
+}
```
Comment (StormBytePP) [inline src/server/game/Handlers/BlackMarketHandler.cpp]: no newline

File: src/server/game/Server/Packets/BlackMarketPackets.cpp (hunk 5 above)
Comment (StormBytePP) [inline src/server/game/Server/Packets/BlackMarketPackets.cpp]: no newline

File: src/server/game/BlackMarket/BlackMarketMgr.h [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+    bool LoadFromDB(Field* fields); // TO-Do implement checks
+}; 
+
+class BlackMarketEntry
+{
+public:
+    int32 MarketId = 0;
+    uint64 CurrentBid = 0;
+    int32 NumBids = 0;
+    ObjectGuid::LowType Bidder = 0;
```
Comment (DDuarte) [inline src/server/game/BlackMarket/BlackMarketMgr.h]: Move these fields to private and add gets and setters to them

File: src/server/game/BlackMarket/BlackMarketMgr.h [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+
+    uint64 GetMinIncrement() const { return (CurrentBid / 20) - ((CurrentBid / 20) % GOLD)  ; } //5% increase every bid (has to be round gold value)
+    bool ValidateBid(uint64 bid) const;
+    void PlaceBid(uint64 bid, Player* player, SQLTransaction& trans);
+
+    std::string BuildAuctionMailSubject(BMAHMailAuctionAnswers response) const;
+    std::string BuildAuctionMailBody();
+
+    void MailSent() { m_mailSent = true; } // Set when mail has been sent
+    bool GetMailSent() { return m_mailSent; }
```
Comment (DDuarte) [inline src/server/game/BlackMarket/BlackMarketMgr.h]: missing const

File: src/server/game/BlackMarket/BlackMarketMgr.h [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+    typedef std::map<int32, BlackMarketEntry*> BlackMarketEntryMap;
+    typedef std::map<int32, BlackMarketTemplate*> BlackMarketTemplateMap;
+
+    // Load templates first
+    void LoadTemplates();
+    void LoadAuctions();
+
+    void Update(bool updateTime = false);
+    void RefreshAuctions();
+    time_t GetLastUpdate() { return mLastUpdate; }
```
Comment (DDuarte) [inline src/server/game/BlackMarket/BlackMarketMgr.h]: missing const

File: src/server/game/BlackMarket/BlackMarketMgr.h [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+    BlackMarketTemplate* GetTemplateByID(int32 marketId) const;
+
+    void AddAuction(BlackMarketEntry* auction);
+    void AddTemplate(BlackMarketTemplate* templ);
+
+    void SendAuctionWonMail(BlackMarketEntry* entry, SQLTransaction& trans);
+    void SendAuctionOutbidMail(BlackMarketEntry* entry, SQLTransaction& trans); // Call before incrementing bid
+
+  private:
+
```
Comment (DDuarte) [inline src/server/game/BlackMarket/BlackMarketMgr.h]: These private members should start with _ (m is not needed)

File: src/server/game/BlackMarket/BlackMarketMgr.h (hunk 2 above)
Comment (DDuarte) [inline src/server/game/BlackMarket/BlackMarketMgr.h]: missing an empty line here

File: src/server/game/Handlers/BankHandler.cpp (hunk 3 above)
Comment (DDuarte) [inline src/server/game/Handlers/BankHandler.cpp]: :)

File: src/server/game/Handlers/BlackMarketHandler.cpp [hunk 11]
```diff
@@ -41,6 +42,113 @@ void WorldSession::SendBlackMarketOpenResult(ObjectGuid guid, Creature* /*auctio
...
+    Creature* unit = GetPlayer()->GetNPCIfCanInteractWith(packet.Guid, UNIT_NPC_FLAG_BLACK_MARKET);
+    if (!unit)
+    {
+        TC_LOG_DEBUG("network", "WORLD: HandleBlackMarketRequestItems - Unit (GUID: %s) not found or you can't interact with him.", packet.Guid.ToString().c_str());
+        return;
+    }
+
+    WorldPackets::BlackMarket::BlackMarketRequestItemsResult result;
+    sBlackMarketMgr->BuildItemsResponse(result, GetPlayer());
+    return SendPacket(result.Write());
```
Comment (DDuarte) [inline src/server/game/Handlers/BlackMarketHandler.cpp]: returning in a void method

File: src/server/game/Server/WorldSession.h [hunk 12]
```diff
@@ -920,6 +925,9 @@ class WorldSession
 
         // Black Market
         void SendBlackMarketOpenResult(ObjectGuid guid, Creature* auctioneer);
+        void SendBlackMarketBidOnItemResult(int32 result, int32 marketId, WorldPackets::Item::ItemInstance* item);
```
Comment (DDuarte) [inline src/server/game/Server/WorldSession.h]: Change ItemInstance* to reference

File: src/server/game/BlackMarket/BlackMarketMgr.cpp [hunk 13]
```diff
@@ -0,0 +1,488 @@
...
+    std::vector<int32> bonusListIDs;
+    for (char const* token : bonusListIDsTok)
+    {
+        int32 bonusListID = atol(token);
+        bonusListIDs.push_back(bonusListID);
+    }
+
+    if (!bonusListIDs.empty())
+    {
+        Item.ItemBonus.emplace();
```
Comment (Vincent-Michael) [inline src/server/game/BlackMarket/BlackMarketMgr.cpp]: no member named 'emplace' in
      'boost::optional<WorldPackets::Item::ItemBonusInstanceData>'
        Item.ItemBonus.emplace();

File: src/server/game/Server/Packets/BlackMarketPackets.cpp (hunk 5 above)
Comment (Gigi1237) [inline src/server/game/Server/Packets/BlackMarketPackets.cpp]: how do you remove newline?
Can't seem to find a way in either VS os NotePad++

Comment (Vincent-Michael) [inline src/server/game/Server/Packets/BlackMarketPackets.cpp]: add new line on end

Comment (Gigi1237) [inline src/server/game/Server/Packets/BlackMarketPackets.cpp]: oh... I thout TC didn't want one.

**Comment by Gigi1237** [comment]
Everything should be fixed now

**Comment by Gigi1237** [comment]
Why the hell did travis fail now???
The build succeded

**Comment by Nyeriah** [comment]
The command "mysql -utrinity -ptrinity hotfixes < sql/base/dev/hotfixes_database.sql" exited with 0.
0.08s$ cat sql/updates/world/*.sql | mysql -utrinity -ptrinity world
ERROR 1146 (42S02) at line 2701: Table 'world.locales_gameobject' doesn't exist

**Comment by Gigi1237** [comment]
@Nyeriah How can I fix it?

**Comment by tkrokli** [comment]
@Gigi1237 : if you look at the error message and its contents, I think I can see what it is pointing at:
```bash
$ cat sql/updates/world/*.sql | mysql -utrinity -ptrinity world
ERROR 1146 (42S02) at line 2701: Table 'world.locales_gameobject' doesn't exist
```
The file `/sql/updates/world/2015_07_18_04_world.sql` references that table, but the table is not in the DB.
Have you tried to update (rebase) your own repo and then push again? Maybe the existing source you are referring to in this PR is outdated and you need to rebase?

File: src/server/game/BlackMarket/BlackMarketMgr.h [hunk 14]
```diff
@@ -0,0 +1,166 @@
...
+      BlackMarketEntryMap mAuctions;
+      BlackMarketTemplateMap mTemplates;
+
+      time_t _LastUpdate;
```
Comment (jackpoz) [inline src/server/game/BlackMarket/BlackMarketMgr.h]: please initialize this field

File: src/server/game/BlackMarket/BlackMarketMgr.h [hunk 15]
```diff
@@ -0,0 +1,166 @@
...
+    void MailSent() { _mailSent = true; } // Set when mail has been sent
+    bool GetMailSent() const { return _mailSent; }
+
+private:
+    int32 _marketId = 0;
+    uint64 _currentBid = 0;
+    int32 _numBids = 0;
+    ObjectGuid::LowType _bidder = 0;
+
+    time_t _secondsRemaining;
```
Comment (jackpoz) [inline src/server/game/BlackMarket/BlackMarketMgr.h]: please initialize this field too just like all the others above

**Comment by Naios** [comment]
Don't forget to update the `updates` table in the fulldump too please.

File: src/server/game/BlackMarket/BlackMarketMgr.cpp [hunk 16]
```diff
@@ -0,0 +1,488 @@
+/*
+* Copyright (C) 2008-2015 TrinityCore <http://www.trinitycore.org/>
```
Comment (Carbenium) [inline src/server/game/BlackMarket/BlackMarketMgr.cpp]: Missing space at the beginning of this and the following header lines.

File: src/server/game/BlackMarket/BlackMarketMgr.cpp [hunk 17]
```diff
@@ -0,0 +1,488 @@
...
+
+    if (!result)
+    {
+        TC_LOG_INFO("server.loading", ">> Loaded 0 black market templates. DB table `blackmarket_template` is empty.");
+        return;
+    }
+
+    uint32 count = 0;
+
+    SQLTransaction trans = WorldDatabase.BeginTransaction();
```
Comment (Carbenium) [inline src/server/game/BlackMarket/BlackMarketMgr.cpp]: I can't really see where this transaction is used.

File: src/server/game/BlackMarket/BlackMarketMgr.cpp [hunk 18]
```diff
@@ -0,0 +1,488 @@
...
+#include "BlackMarketMgr.h"
+#include "Player.h"
+#include "Language.h"
+
+void BlackMarketMgr::LoadTemplates()
+{
+    uint32 oldMSTime = getMSTime();
+
+    // Clear in case we are reloading
+    if (!mTemplates.empty())
```
Comment (Carbenium) [inline src/server/game/BlackMarket/BlackMarketMgr.cpp]: According to out codestyle guide local variables should be lower case without a prefix.

File: src/server/game/BlackMarket/BlackMarketMgr.cpp [hunk 19]
```diff
@@ -0,0 +1,488 @@
...
+
+    TC_LOG_INFO("server.loading", ">> Loaded %u black market templates in %u ms.", count, GetMSTimeDiffToNow(oldMSTime));
+}
+
+void BlackMarketMgr::LoadAuctions()
+{
+    uint32 oldMSTime = getMSTime();
+
+    // Clear in case we are reloading
+    if (!mAuctions.empty())
```
Comment (Carbenium) [inline src/server/game/BlackMarket/BlackMarketMgr.cpp]: Same here (prefix).

File: src/server/game/BlackMarket/BlackMarketMgr.h [hunk 20]
```diff
@@ -0,0 +1,166 @@
+/*
+* Copyright (C) 2008-2015 TrinityCore <http://www.trinitycore.org/>
```
Comment (tkrokli) [inline src/server/game/BlackMarket/BlackMarketMgr.h]: Like Carbenium said: "Missing space at the beginning of this and the following header lines."
(Lines 2 - 16 need 1 space in front of them to follow the source standard.)

File: src/server/game/BlackMarket/BlackMarketMgr.cpp [hunk 21]
```diff
@@ -0,0 +1,488 @@
...
+
+    // Log trade
+    if (logGmTrade)
+    {
+        sLog->outCommand(bidderAccId, "GM %s (Account: %u) won item in blackmarket auction: %s (Entry: %u Count: %u) and payed gold : %u.",
+            bidderName.c_str(), bidderAccId, item->GetTemplate()->GetDefaultLocaleName(), item->GetEntry(), item->GetCount(), entry->GetCurrentBid() / GOLD);
+    }
+
+    if (bidder)
+    {
```
Comment (Carbenium) [inline src/server/game/BlackMarket/BlackMarketMgr.cpp]: Braces can be omitted.

**Comment by Aokromes** [comment]
This branch has conflicts that must be resolved
 Use the command line to resolve conflicts before continuing.
+Carbenium, jackpoz, Naios comments.

**Comment by Ascathor** [comment]
Any update on this @Gigi1237 ?

**Comment by Gigi1237** [comment]
@Ascathor haven't had time to update and fix this. Life gets in the way. I'll see if I can do it in the next month.

**Comment by Ascathor** [comment]
@Gigi1237 Fair enough. Just curious

**Comment by msoky** [comment]
any news?

**Comment by Shauren** [comment]
Merged: c19172cad9a3302b7f5c8d5ca1d96de9a7311847

**Comment by msoky** [comment]
thank you, I want only point out to old pull requests..

**Comment by Gigi1237** [comment]
@Shauren Thanks for merging. Sorry for not being able to update it. I don't have the time I once had to work on projects for fun :disappointed:

**Pull request closed without merging**